src/Workspaces/Core/Portable/Workspace/Solution/SourceGeneratedDocumentState.cs

```diff
@@ -69,31 +97,33 @@ internal sealed class SourceGeneratedDocumentState : DocumentState
         ConstantTextAndVersionSource textSource,
         SourceText text,
         LoadTextOptions loadTextOptions,
-        AsyncLazy<TreeAndVersion> treeSource)
+        AsyncLazy<TreeAndVersion> treeSource,
+        Lazy<Checksum> lazyTextChecksum)
         : base(languageServices, documentServiceProvider, attributes, options, textSource, loadTextOptions, treeSource)
     {
         Identity = documentIdentity;
 
         SourceText = text;
-
-        _lazyTextChecksum = new Lazy<Checksum>(() => Checksum.From(SourceText.GetChecksum()));
+        _lazyTextChecksum = lazyTextChecksum;
     }
 
+    private static Checksum ComputeChecksum(SourceText text)
+        => Checksum.From(text.GetChecksum());
```

Comment:
Nit: I almost don't like this method since it hides that text.GetChecksum() is being called (versus whether this is some other checksum approach.) But no reason to touch this since I suspect this will change per our conversation and your upcoming proposal.